File: android/app/src/main/java/com/pumplo/app/InstagramSharePlugin.java
package com.pumplo.app;

import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;

import androidx.core.content.FileProvider;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.io.File;

// Shares a finished-workout card directly as an Instagram Story BACKGROUND via
// the ADD_TO_STORY intent. The generic share sheet hands IG a file, which IG
// drops onto the live camera as a sticker; using ADD_TO_STORY with the image as
// the intent data makes it the full Story background instead.
// Needs <queries> visibility for com.instagram.android (AndroidManifest).
@CapacitorPlugin(name = "InstagramShare")
public class InstagramSharePlugin extends Plugin {

    private static final String IG_PACKAGE = "com.instagram.android";

    @PluginMethod
    public void isInstalled(PluginCall call) {
        boolean installed;
        try {
            getContext().getPackageManager().getPackageInfo(IG_PACKAGE, 0);
            installed = true;
        } catch (PackageManager.NameNotFoundException e) {
            installed = false;
        }
        JSObject ret = new JSObject();
        ret.put("installed", installed);
        call.resolve(ret);
    }

    @PluginMethod
    public void shareToStories(PluginCall call) {
        String path = call.getString("imagePath");
        if (path == null) {
            call.reject("missing imagePath");
            return;
        }
        try {
            Uri parsed = Uri.parse(path);
            File file = new File(parsed.getPath());
            Uri contentUri = FileProvider.getUriForFile(
                getContext(),
                getContext().getPackageName() + ".fileprovider",
                file
            );

            Intent intent = new Intent("com.instagram.share.ADD_TO_STORY");
            intent.setDataAndType(contentUri, "image/jpeg");
            intent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
            intent.putExtra("source_application", getContext().getPackageName());

            getContext().grantUriPermission(
                IG_PACKAGE, contentUri, Intent.FLAG_GRANT_READ_URI_PERMISSION);

            if (intent.resolveActivity(getContext().getPackageManager()) != null) {
                getActivity().startActivityForResult(intent, 0);
                call.resolve();
            } else {
                call.reject("instagram not installed");
            }
        } catch (Exception e) {
            call.reject("share failed: " + e.getMessage());
        }
    }
}

File: ios/App/App/InstagramSharePlugin.swift
import Foundation
import Capacitor
import UIKit

// Shares a finished-workout card directly as an Instagram Story BACKGROUND.
// The generic share sheet hands IG a file, which IG drops onto the live camera
// as a sticker (you see the selfie behind the card). Instead we use IG's
// documented Stories handoff: put the PNG on the pasteboard under the
// `com.instagram.sharedSticker.backgroundImage` key and open the
// `instagram-stories://share` URL, which makes the image the full background.
// Requires `instagram-stories` in LSApplicationQueriesSchemes (Info.plist).
@objc(InstagramSharePlugin)
public class InstagramSharePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "InstagramSharePlugin"
    public let jsName = "InstagramShare"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "isInstalled", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "shareToStories", returnType: CAPPluginReturnPromise)
    ]

    private static let storiesScheme = "instagram-stories://share?source_application=com.pumplo.app"

    @objc func isInstalled(_ call: CAPPluginCall) {
        guard let url = URL(string: InstagramSharePlugin.storiesScheme) else {
            call.resolve(["installed": false]); return
        }
        DispatchQueue.main.async {
            call.resolve(["installed": UIApplication.shared.canOpenURL(url)])
        }
    }

    @objc func shareToStories(_ call: CAPPluginCall) {
        guard let path = call.getString("imagePath") else {
            call.reject("missing imagePath"); return
        }
        // Filesystem returns a file:// URI; fall back to a bare path if needed.
        let fileURL = URL(string: path) ?? URL(fileURLWithPath: path)
        guard let data = try? Data(contentsOf: fileURL) else {
            call.reject("could not read image at \(path)"); return
        }
        guard let storiesURL = URL(string: InstagramSharePlugin.storiesScheme) else {
            call.reject("bad instagram url"); return
        }

        DispatchQueue.main.async {
            guard UIApplication.shared.canOpenURL(storiesURL) else {
                call.reject("instagram not installed"); return
            }
            let items: [[String: Any]] = [["com.instagram.sharedSticker.backgroundImage": data]]
            let options: [UIPasteboard.OptionsKey: Any] = [
                .expirationDate: Date().addingTimeInterval(60 * 5)
            ]
            UIPasteboard.general.setItems(items, options: options)
            UIApplication.shared.open(storiesURL, options: [:]) { success in
                if success { call.resolve() } else { call.reject("failed to open instagram") }
            }
        }
    }
}
